core/src/main/java/org/apache/carbondata/core/datamap/StreamDataMap.java

```diff
@@ -0,0 +1,148 @@
+    this.carbonTable = carbonTable;
+  }
+
+  public void init(FilterResolverIntf filterExp) {
+    if (filterExp != null) {
+      // cache all columns
+      List<CarbonColumn> minMaxCacheColumns = new ArrayList<>();
+      for (CarbonDimension dimension : carbonTable.getDimensions()) {
+        if (!dimension.isComplex()) {
+          minMaxCacheColumns.add(dimension);
+        }
+      }
+      minMaxCacheColumns.addAll(carbonTable.getMeasures());
+      // prepare cardinality of all dimensions
+      List<ColumnSchema> listOfColumns =
+          carbonTable.getTableInfo().getFactTable().getListOfColumns();
+      int[] columnCardinality = new int[listOfColumns.size()];
+      for (int index = 0; index < columnCardinality.length; index++) {
+        columnCardinality[index] = Integer.MAX_VALUE;
+      }
+      // initial filter executor
+      SegmentProperties segmentProperties =
+          new SegmentProperties(listOfColumns, columnCardinality);
+      filterExecuter = FilterUtil.getFilterExecuterTree(
+          filterExp, segmentProperties, null, minMaxCacheColumns);
+    }
+  }
+
+  public List<StreamFile> prune(List<Segment> segments) throws IOException {
+    if (filterExecuter == null) {
+      // if filter is null, list all steam files
+      return listAllStreamFiles(segments, false);
+    } else {
+      List<StreamFile> streamFileList = new ArrayList<>();
+      for (StreamFile streamFile : listAllStreamFiles(segments, true)) {
+        if (isScanRequire(streamFile)) {
+          // if stream file is required to scan
+          streamFileList.add(streamFile);
+          streamFile.setMinMaxIndex(null);
+        }
+      }
+      return streamFileList;
+    }
+  }
+
+  private boolean isScanRequire(StreamFile streamFile) {
+    // backward compatibility, old stream file without min/max index
+    if (streamFile.getMinMaxIndex() == null) {
+      return true;
+    }
+    byte[][] maxValue = streamFile.getMinMaxIndex().getMaxValues();
+    byte[][] minValue = streamFile.getMinMaxIndex().getMinValues();
+    BitSet bitSet = filterExecuter.isScanRequired(maxValue, minValue);
+    if (!bitSet.isEmpty()) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // TODO optimize and move the code to StreamSegment , but it's in the streaming module.
+  private List<StreamFile> listAllStreamFiles(List<Segment> segments, boolean withMinMax)
+      throws IOException {
+    List<StreamFile> streamFileList = new ArrayList<>();
+    for (Segment segment : segments) {
+      String segmentDir = CarbonTablePath.getSegmentPath(
+          carbonTable.getAbsoluteTableIdentifier().getTablePath(), segment.getSegmentNo());
+      FileFactory.FileType fileType = FileFactory.getFileType(segmentDir);
+      if (FileFactory.isFileExist(segmentDir, fileType)) {
+        SegmentIndexFileStore segmentIndexFileStore = new SegmentIndexFileStore();
+        segmentIndexFileStore.readAllIIndexOfSegment(segmentDir);
```

Comment:
It involves the listing of a folder and gets the index file. Its a costly operation for s3 file system as you are not even caching it. So it is better fix the carbonindex file name in case of streaming and read it directly instead of listing.

Reply:
fixed